Error when reinstalling vscode

### Expected behavior

As a user, I want to be able to reinstall vscode without any errors.

### Actual behavior

Reinstalling threw exception.

```
Start: Install vscode
Due to lack of permissions, Microsoft's mklink with junction had to be used to create a Symlink. See https://github.com/devonfw/IDEasy/blob/main/documentation/symlinks.asciidoc for further details. Error was: C:\Users\<username>\projects\IDEasy\software\vscode: Dem Client fehlt ein erforderliches Recht
You are on Windows and you do not have permissions to create symbolic links. Junctions are used as an alternative, however, these can not point to relative paths. So the source (..\..\_ide\software\default\vscode\vscode\1.100.1) is interpreted as an absolute path.
Verbindung erstellt für C:\Users\<username>\projects\IDEasy\software\vscode <<===>> C:\Users\<username>\projects\_ide\software\default\vscode\vscode\1.100.1
Step 'ide (install,vscode)' failed: java.lang.IllegalStateException: Failed to delete C:\Users\<username>\projects\IDEasy\plugins\vscode
 Step 'Install vscode (*)' failed: java.lang.IllegalStateException: Failed to delete C:\Users\<username>\projects\IDEasy\plugins\vscode
2 step(s) failed out of 2 steps.
An unexpected error occurred!
We are sorry for the inconvenience.
Please check the error below, resolve it and try again.
If the error is not on your end (network connectivity, lack of permissions, etc.) please file a bug:
https://github.com/devonfw/IDEasy/issues/new?assignees=&labels=bug&projects=&template=bug.md&title=IllegalStateException%3A+Failed+to+delete+C%3A%5CUsers%5C<username>%5Cprojects%5CIDEasy%5Cplugins%5Cvscode
java.lang.IllegalStateException: Failed to delete C:\Users\<username>\projects\IDEasy\plugins\vscode
        at com.devonfw.tools.ide.io.FileAccessImpl.delete(FileAccessImpl.java:791)
        at com.devonfw.tools.ide.tool.plugin.PluginBasedCommandlet.postInstall(PluginBasedCommandlet.java:104)
        at com.devonfw.tools.ide.tool.LocalToolCommandlet.install(LocalToolCommandlet.java:98)
        at com.devonfw.tools.ide.tool.ToolCommandlet.install(ToolCommandlet.java:260)
        at com.devonfw.tools.ide.commandlet.InstallCommandlet.run(InstallCommandlet.java:77)
        at com.devonfw.tools.ide.context.AbstractIdeContext.applyAndRun(AbstractIdeContext.java:976)
        at com.devonfw.tools.ide.context.AbstractIdeContext.run(AbstractIdeContext.java:894)
        at com.devonfw.tools.ide.cli.Ideasy.runOrThrow(Ideasy.java:95)
        at com.devonfw.tools.ide.cli.Ideasy.run(Ideasy.java:59)
        at com.devonfw.tools.ide.cli.Ideasy.main(Ideasy.java:133)
        at java.base@21.0.2/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.nio.file.AccessDeniedException: C:\Users\<username>\projects\IDEasy\plugins\vscode\angular.ng-template-20.0.0\angular.png
        at java.base@21.0.2/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:273)
        at java.base@21.0.2/sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:104)
        at java.base@21.0.2/java.nio.file.Files.delete(Files.java:1152)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:807)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:802)
        at com.devonfw.tools.ide.io.FileAccessImpl.deleteRecursive(FileAccessImpl.java:802)
        at com.devonfw.tools.ide.io.FileAccessImpl.delete(FileAccessImpl.java:788)
        ... 10 more


Error: IDEasy failed with exit code 255
```

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. install vscode
2. uninstall vscode
3. install vscode again

### Related/Dependent Issues


### Comments/Hints:
Same error was thrown on two team members computers as well.
Workaround:
* manually delete angular.png

Further Finding:
* After deleting angular.png and executing the steps to reproduce the error, (install, uninstall, install) another file caused the same error: ...\projects\IDEasy\plugins\vscode\angular.ng-template-20.0.0\index.js

### Affected version:

* OS: Windows 11
* IDEasy: 2025.04.002
* VSCode: 1.100.1
* VSCode Plugin causing the error: angular.ng-template-20.0.0

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -2,6 +2,14 @@
 
 This file documents all notable changes to https://github.com/devonfw/IDEasy[IDEasy].
 
+== 2025.06.002
+
+Release with new features and bugfixes:
+
+* https://github.com/devonfw/IDEasy/issues/1304[#1304]: Error when reinstalling vscode
+
+The full list of changes for this release can be found in https://github.com/devonfw/IDEasy/milestone/29?closed=1[milestone 2025.06.002].
+
 == 2025.06.001
 
 Release with new features and bugfixes:

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
@@ -292,7 +292,7 @@ public interface FileAccess {
    * @param writable {@code true} to make the file writable, {@code false} to make it read-only
    * @return {@code true} if the operation was successful or supported, {@code false} otherwise
    */
-  boolean setWriteable(Path file, boolean writable);
+  boolean setWritable(Path file, boolean writable);
 
   /**
    * Makes a file executable (analog to 'chmod a+x').

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
@@ -286,6 +286,15 @@ public interface FileAccess {
   boolean isEmptyDir(Path dir);
 
   /**
+   * Sets or unsets the writable permission for the specified file path.
+   *
+   * @param file {@link Path} to the file.
+   * @param writable {@code true} to make the file writable, {@code false} to make it read-only
+   * @return {@code true} if the operation was successful or supported, {@code false} otherwise
+   */
+  boolean setWriteable(Path file, boolean writable);
+
+  /**
    * Makes a file executable (analog to 'chmod a+x').
    *
    * @param file {@link Path} to the file.

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
@@ -804,7 +804,10 @@ public class FileAccessImpl implements FileAccess {
       }
     }
     this.context.trace("Deleting {} ...", path);
-    path.toFile().setWritable(true);
+    boolean isSetWritable = path.toFile().setWritable(true);
+    if (!isSetWritable) {
+      this.context.debug("Couldn't give write access to file: " + path);
+    }
     Files.delete(path);
   }
 

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
@@ -22,7 +22,9 @@ import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.DosFileAttributeView;
 import java.nio.file.attribute.FileTime;
+import java.nio.file.attribute.PosixFileAttributeView;
 import java.nio.file.attribute.PosixFilePermission;
 import java.nio.file.attribute.PosixFilePermissions;
 import java.security.DigestInputStream;
@@ -820,7 +822,7 @@ public class FileAccessImpl implements FileAccess {
       }
     }
     this.context.trace("Deleting {} ...", path);
-    boolean isSetWritable = path.toFile().setWritable(true);
+    boolean isSetWritable = setWriteable(path, true);
     if (!isSetWritable) {
       this.context.debug("Couldn't give write access to file: " + path);
     }
@@ -946,6 +948,38 @@ public class FileAccessImpl implements FileAccess {
       }
     }
     return null;
+  }
+
+  @Override
+  public boolean setWriteable(Path file, boolean writeable) {
+    try {
+      PosixFileAttributeView posix = Files.getFileAttributeView(file, PosixFileAttributeView.class);
+
+      if (posix != null) {
+        // POSIX
+        Set<PosixFilePermission> permissions = new HashSet<>(posix.readAttributes().permissions());
+        boolean changed;
+        if (writeable) {
+          changed = permissions.add(PosixFilePermission.OWNER_WRITE);
+        } else {
+          changed = permissions.remove(PosixFilePermission.OWNER_WRITE);
+        }
+        if (changed) {
+          posix.setPermissions(permissions);
+        }
+        return true;
+
+      } else {
+        // Windows
+        DosFileAttributeView dos = Files.getFileAttributeView(file, DosFileAttributeView.class);
+        if (dos != null) {
+          dos.setReadOnly(!writeable);
+        }
+        return true;
+      }
+    } catch (IOException e) {
+      return false;
+    }
   }
 
   @Override

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
@@ -822,7 +822,7 @@ public class FileAccessImpl implements FileAccess {
       }
     }
     this.context.trace("Deleting {} ...", path);
-    boolean isSetWritable = setWriteable(path, true);
+    boolean isSetWritable = setWritable(path, true);
     if (!isSetWritable) {
       this.context.debug("Couldn't give write access to file: " + path);
     }
@@ -951,14 +951,14 @@ public class FileAccessImpl implements FileAccess {
   }
 
   @Override
-  public boolean setWriteable(Path file, boolean writeable) {
+  public boolean setWritable(Path file, boolean writable) {
     try {
       // POSIX
       PosixFileAttributeView posix = Files.getFileAttributeView(file, PosixFileAttributeView.class);
       if (posix != null) {
         Set<PosixFilePermission> permissions = new HashSet<>(posix.readAttributes().permissions());
         boolean changed;
-        if (writeable) {
+        if (writable) {
           changed = permissions.add(PosixFilePermission.OWNER_WRITE);
         } else {
           changed = permissions.remove(PosixFilePermission.OWNER_WRITE);
@@ -972,13 +972,15 @@ public class FileAccessImpl implements FileAccess {
       // Windows
       DosFileAttributeView dos = Files.getFileAttributeView(file, DosFileAttributeView.class);
       if (dos != null) {
-        dos.setReadOnly(!writeable);
+        dos.setReadOnly(!writable);
         return true;
       }
 
+      this.context.debug("Failed to set writing permission for file {}", file);
       return false;
 
     } catch (IOException e) {
+      this.context.debug("Error occurred when trying to set writing permission for file " + file + ": " + e);
       return false;
     }
   }

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
@@ -804,6 +804,7 @@ public class FileAccessImpl implements FileAccess {
       }
     }
     this.context.trace("Deleting {} ...", path);
+    path.toFile().setWritable(true);
     Files.delete(path);
   }
 

--- a/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
+++ b/cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
@@ -953,10 +953,9 @@ public class FileAccessImpl implements FileAccess {
   @Override
   public boolean setWriteable(Path file, boolean writeable) {
     try {
+      // POSIX
       PosixFileAttributeView posix = Files.getFileAttributeView(file, PosixFileAttributeView.class);
-
       if (posix != null) {
-        // POSIX
         Set<PosixFilePermission> permissions = new HashSet<>(posix.readAttributes().permissions());
         boolean changed;
         if (writeable) {
@@ -968,15 +967,17 @@ public class FileAccessImpl implements FileAccess {
           posix.setPermissions(permissions);
         }
         return true;
+      }
 
-      } else {
-        // Windows
-        DosFileAttributeView dos = Files.getFileAttributeView(file, DosFileAttributeView.class);
-        if (dos != null) {
-          dos.setReadOnly(!writeable);
-        }
+      // Windows
+      DosFileAttributeView dos = Files.getFileAttributeView(file, DosFileAttributeView.class);
+      if (dos != null) {
+        dos.setReadOnly(!writeable);
         return true;
       }
+
+      return false;
+
     } catch (IOException e) {
       return false;
     }

